### ISSUE TITLE
CentOS7/PHP7 failing @ "Install Drupal with drush" with "RuntimeException: Unable to use ApcClassLoader as APC is not enabled."

### ISSUE THREAD

**Issue by illepic** [geerlingguy/drupal-vm#362]
Quick check: should config.yml declare APC explicitly anywhere?

After running (on OSX):

    sudo ansible-galaxy install -r provisioning/requirements.yml --force

and then following this [CentOS7/PHP7 blog post](http://www.midwesternmac.com/blogs/jeff-geerling/drupal-8-redis-php-7-nginx-and), my build now errors out here:

```shell
TASK: [Generate Drupal site with drush makefile.] *****************************
skipping: [downfalldev]

TASK: [Check if site is already installed.] ***********************************
ok: [downfalldev]

TASK: [Install Drupal with drush.] ********************************************
failed: [downfalldev] => {"changed": true, "cmd": ["/usr/local/bin/drush", "site-install", "standard", "-y", "--site-name=DF", "--account-name=admin", "--account-pass=admin", "--db-url=mysql://dfdbuser:dfdbpass@localhost/downfall_d8"], "delta": "0:00:00.497109", "end": "2015-12-28 17:31:43.782567", "rc": 1, "start": "2015-12-28 17:31:43.285458", "warnings": []}
stderr: Starting Drupal installation. This takes a while. Consider using the [ok]
--notify global option.
RuntimeException: Unable to use ApcClassLoader as APC is not enabled.[error]
in
/var/www/df/web/d8/vendor/symfony/class-loader/ApcClassLoader.php:71
Stack trace:
#0 /var/www/df/web/d8/core/lib/Drupal/Core/DrupalKernel.php(977):
Symfony\Component\ClassLoader\ApcClassLoader->__construct('drupal.class_lo...',
Object(Composer\Autoload\ClassLoader))
#1 /var/www/df/web/d8/core/lib/Drupal/Core/DrupalKernel.php(264):
Drupal\Core\DrupalKernel->initializeSettings(Object(Symfony\Component\HttpFoundation\Request))
#2 /var/www/df/web/d8/core/includes/install.core.inc(403):
Drupal\Core\DrupalKernel::createFromRequest(Object(Symfony\Component\HttpFoundation\Request),
Object(Composer\Autoload\ClassLoader), 'install', true)
#3 /var/www/df/web/d8/core/includes/install.core.inc(113):
install_begin_request(Object(Composer\Autoload\ClassLoader), Array)
#4 /usr/local/share/drush/includes/drush.inc(725):
install_drupal(Object(Composer\Autoload\ClassLoader), Array)
#5 /usr/local/share/drush/includes/drush.inc(710):
drush_call_user_func_array('install_drupal', Array)
#6 /usr/local/share/drush/commands/core/drupal/site_install.inc(80):
drush_op('install_drupal', Object(Composer\Autoload\ClassLoader),
Array)
#7 /usr/local/share/drush/commands/core/site_install.drush.inc(247):
drush_core_site_install_version('standard', Array)
#8 /usr/local/share/drush/includes/command.inc(366):
drush_core_site_install('standard')
#9 /usr/local/share/drush/includes/command.inc(217):
_drush_invoke_hooks(Array, Array)
#10 /usr/local/share/drush/includes/command.inc(185):
drush_command('standard')
#11 /usr/local/share/drush/lib/Drush/Boot/BaseBoot.php(67):
drush_dispatch(Array)
#12 /usr/local/share/drush/includes/preflight.inc(66):
Drush\Boot\BaseBoot->bootstrap_and_dispatch()
#13 /usr/local/share/drush/drush.php(12): drush_main()
#14 {main}
stdout: You are about to DROP all tables in your 'downfall_d8' database. Do you want to continue? (y/n): y

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/Users/illepic/playbook.retry

downfalldev                : ok=120  changed=60   unreachable=0    failed=1

Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
```

My `config.yml` [can be found here](https://gist.github.com/illepic/f89ce9a823ebde1909a1). Did an udpated Ansible role change the inclusion of APC?

**Comment by geerlingguy** [comment]
Seems odd... APCu should be available. Did you follow [these directions](http://docs.drupalvm.com/en/latest/other/php-7/#redhatcentos-7) from the docs and make sure all of that is configured correctly?

`php-pecl-apc` is installed by default with the config.yml you're using, as it's inherited from the defaults in the PHP role: https://github.com/geerlingguy/ansible-role-php/blob/master/vars/RedHat.yml#L15

**Comment by geerlingguy** [comment]
@illepic - Can you log in (`vagrant ssh`) and run `php -m` and give the output?

**Comment by illepic** [comment]
Here is the output from `php -m`:

```shell
[vagrant@downfall ~]$ php -m
[PHP Modules]
apcu
bz2
calendar
Core
ctype
curl
date
dom
exif
fileinfo
filter
ftp
gd
gettext
hash
iconv
imap
json
ldap
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
readline
redis
Reflection
session
shmop
SimpleXML
sockets
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tokenizer
wddx
xml
xmlreader
xmlrpc
xmlwriter
xsl
Zend OPcache
zlib

[Zend Modules]
Zend OPcache
```

I should note I'm on OSX, Vagrant 1.8.1, Virtualbox 5.0.10. (No, I'm not always That Windows Guy :P )

**Comment by geerlingguy** [comment]
It looks like `apcu` is loaded, and that provides the same backwards-compatible API AFAICT.

However, maybe it's just that PHP needs a restart. Can you run `sudo service php-fpm restart` and see if that helps? Also try a full destroy and `vagrant up` again.

> I should note I'm on OSX, Vagrant 1.8.1, Virtualbox 5.0.10. (No, I'm not always That Windows Guy :P )

Haha, I had my trigger finger on that 'windows' label just in case :)

**Comment by illepic** [comment]
Fun fact: make sure that `drupal.make.yml` sets the Drupal core version to something stable or you end up on the active dev branch. (I knew this, really I did!)

So, here is how drupal.make.yml should actually look:

```yaml
---
api: 2

# Basic Drush Make file for Drupal. Be sure to update the drupal_major_version
# variable inside config.yml if you change the major version in this file.

# Drupal core (major version, e.g. 6.x, 7.x, 8.x).
core: "8.x"

defaults:
  projects:
    subdir: "contrib"

projects:

  # Core.
  drupal:
    type: "core"
    download:
      # Drupal core branch (e.g. "6.x", "7.x", "8.0.x").
      # branch: "8.0.x" CHANGED TO BELOW
      branch: "8.0.1"
      working-copy: true

  # Other modules.
  devel: "1.x-dev"
  redis:
    download:
      type: git
      url: https://github.com/md-systems/redis.git
      branch: 8.x-1.x
  migrate_upgrade: "1.x-dev"
  migrate_plus: "1.x-dev"
  migrate_tools: "1.x-dev"
```

Apparently 8.0.x branch of Drupal is doing something funny around APC (today? recently?), whereas 8.0.1 is nice and stable.

Thanks for letting me bug you again!

**Comment by geerlingguy** [comment]
Ah, I see... interesting to note! I know Drupal switched something with APC a few months back, but not anything more recent. D'oh!

**Comment by illepic** [comment]
Blew away my home Windows machine because I was tired of Windows derping on everything. Moved to Ubuntu Desktop 15.10. Fired up a fresh clone of Drupal VM, same `drupal.make.yml` as above (containing `branch: "8.0.1"` mentioned above).

Ended up with the same error:

    RuntimeException: Unable to use ApcClassLoader as APC is not enabled.

So, I ssh'd into the box, ran `sudo service php-fpm restart && rm -rf /var/www/path/to/mapped/site` to restart php-fpm and delete all drupal files. I exited out of the box and then ran `vagrant provision` to cause a full drush reinstall of drupal 8.

Success. I think the issue here is that from a fully fresh `vagrant up` there needs to be a php-fpm restart before the drush make happens. Since I experienced this fresh on both OSX and Ubuntu, I'm pretty sure this is actually a thing. (My resolution on OSX above did include me restarting php-fpm before changing the Drupal version to 8.0.1)

**Comment by geerlingguy** [comment]
Strange... I'll have to do a `vagrant up` with your config.yml and see if I can figure out what's going on. Is your drupal.make.yml file customized at all? Can you post a gist of it as well?

**Comment by geerlingguy** [comment]
Ah, nevermind, saw your makefile above. Something in there might be the culprit too... I'll try as soon as I get a chance.

**Comment by geerlingguy** [comment]
Oh... just noticed this is with nginx, which is using `php-fpm`, which means you're correct in that it probably just needs a reload before the drupal installation begins... so I'll attack it from that angle and see if I can (a) reproduce the failure you're seeing, and (b) fix it :)

Hopefully a simple fix.

**Comment by geerlingguy** [comment]
Hmm... completely destroyed and rebuilt my VM twice now—once with stock config (Apache) and your `drupal.make.yml` and once with the exact `config.yml` and `drupal.make.yml` you've linked... I can't get anything to fail, unfortunately :/

It seems that maybe one of the roles is out of date on your host? That may cause the problem—can you run `ansible-galaxy install -r provisioning/requirements.yml --force`?

**Comment by gapple** [comment]
Came here searching for the same problem in my own environment, thought I would comment on the cause.

The change that causes this issue is in 8.0.2
`DrupalKernel::initializeSettings()` was changed from checking `function_exists('apc_fetch')` to `function_exists('apcu_fetch')`.  However `ApcClassLoader::__construct` only checks `extension_loaded('apc')`.
The [APCu Installation Docs](https://secure.php.net/manual/en/apcu.installation.php) note that PHP7 requires a separate extension for backwards compatibility, instead of the previous `--enable-apcu-bc` configuration option.

I added a patch to https://www.drupal.org/node/2646100 to fix the check to avoid the exception, but the BC layer is still required to actually make use of Symfony's ApcClassLoader

**Comment by geerlingguy** [comment]
Since this is fixed and will be in Drupal 8.0.3, I'm going to close this issue and point others to it if anyone else notes the problem between now and 8.0.3's release.
